src/weights.jl

```diff
@@ -2,44 +2,189 @@
 ###### Weight vector #####
 
 if VERSION < v"0.6.0-dev.2123"
-    immutable WeightVec{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
-        values::V
-        sum::S
-    end
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:RealVector} <: RealVector{T} end
 else
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
-        values::V
-        sum::S
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T} end
+end
+
+"""
+    @weights name
+
+generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
+and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
+"""
+macro weights(name)
+    return quote
+        if VERSION < v"0.6.0-dev.2123"
+            immutable $name{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        else
+            immutable $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        end
     end
 end
 
+eltype(wv::AbstractWeights) = eltype(wv.values)
+length(wv::AbstractWeights) = length(wv.values)
+values(wv::AbstractWeights) = wv.values
+sum(wv::AbstractWeights) = wv.sum
+isempty(wv::AbstractWeights) = isempty(wv.values)
+
+Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
+Base.size(wv::AbstractWeights) = size(wv.values)
+
+"""
+    cfactor(n::Integer, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise it will return ``\\frac{1}{n}``.
+"""
+cfactor(n::Integer, corrected=false) = 1 / (n - Int(corrected))
+
+"""
+    cfactor(wv::AbstractWeights, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with a set of
+weights `wv`.
+"""
+cfactor(wv::AbstractWeights, corrected=false) = cfactor(wv, Val{corrected})
+
+"""
+    cfactor(wv::AbstractWeights, false)
+
+``\\frac{1}{\sum w}``
+"""
+cfactor(wv::AbstractWeights, ::Type{Val{false}}) = 1 / sum(wv)
+cfactor(wv::AbstractWeights, ::Type{Val{true}}) =
+    throw(ArgumentError("$(typeof(wv)) does not support bias correction."))
+
+
+@weights AnalyticWeights
+
+"""
+    AnalyticWeights(vs, wsum=sum(vs))
+
+Construct an `AnalyticWeights` vector with weight values `vs` and sum of weights `wsum`.
+
+Analytic weights describe a non-random relative importance (usually between 0 and 1)
```

Comment:
> proportional to the inverse of the observation's variance, which typically arises when a value is the average of several observations

I don't know if that's always true though. For example, I think analytic or reliability weights can also be applied for arbitrary weight schemas (e.g., exponential weighting discussed below). The reason for the "non-random" is that these weights assume that there may be some external structure to the weighting (unlike frequency weights); this structure could be the exponential weighting or the inverse variance of each mean observation.